### DISCUSSION TITLE
Mould King Protocoll

### DISCUSSION THREAD

**Discussion by KrustyFPV** [pascallanger/DIY-Multiprotocol-TX-Module#569] [General]
Hi Pascal, 
first of all thank u for your great Work.

i have some Technic Brick models (similar to LEGO) wich are based on 2.4 ghz technic.
for example:
https://tradingshenzhen.com/de/toys-klemmbausteine/mould-king-13112-mechanischer-bagger-1830-teile

do u know about them and can tell me wich Protocol i can use or can implement a Protocol for these?

Since your name sound German i think it should not be a problem to send u an RX TX and Motor.

regards Krusty

**Comment by pascallanger** [comment]
Hi Krusty,
Can you open the TX to see which RF component they are using? taking a good picture (where you can read the components marking) of the area nearby where the antenna is connected.
FYI, I'm based in France if shipping is needed.
Regards, Pascal

**Reply by KrustyFPV** [reply]
Hi Pascal,
https://ibb.co/5r2tGS4 
Unfortunately no text on the IC can i Measure something to specify it.
Shipping France also no Problem Thanks a Lot

Regards
Krusty

**Reply by pascallanger** [reply]
Is it the only chip on the board?
In any case I would need access to the hardware like you propose but there is no guarantee of success.
If you are willing to ship it then consider the cost of the return as well. If ok then contact me by PM on RCGroups username hpnuts.

**Reply by KrustyFPV** [reply]
Hi Pascal,

here is a bit more information i think this is the IC

https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwju3O3QipDwAhVhgf0HHbBmA3wQFjAAegQIBRAD&url=http%3A%2F%2Fwww.quartz1.com%2Fprice%2FPIC%2F410N0629616.pdf&usg=AOvVaw0I5EHyyev8PJAxPbPaQzin

i will make an account at RC group tomorrow have to go to Drink beer now

Regards

**Reply by KrustyFPV** [reply]
Hi Pascal,
i dont recieved the rcgroups Activation e-Mail yet and yes this is the only chip on the board.
i contact u as soon as RC Groups account is working.

**Comment by pascallanger** [comment]
The protocol is implemented in v1.3.2.82 available here: https://downloads.multi-module.org/latest-test/
The doc: https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/Protocols_Details.md#mouldkg---90
Please test and report.

**Comment by marcrupprath** [comment]
I am also interessted in using the mould king protokoll
And i am unfamiliar with the multi module.
Question is how to test this specific protocoll with multi module and 2 Mouldking boxes.

I dont't know how two setup the multi module and how two select mould king protokoll.

I have take a look into source of:

mouldking_NRF24L01,ino and got completely confused:

 - How two start
 - How two adressa specific channel of a box, for example two change motor direktion and speed

Thanks in advance

Marc

**Reply by pascallanger** [reply]
Do you have a radio with a multi module?
The documentation on how to use this protocol is here: https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/Protocols_Details.md#mouldkg---90

**Comment by marcrupprath** [comment]
Hello,
i have this this transmitter RadioMaster-TX16S-Mark-II with  4-IN-1
and in addition this modue: iRangeX IRX4 Plus 2.4G, mainly for tests,

I don't understand waht ist meant with : 
"To associate a brick to a RX number (RX_num above), set this RX number under the protocol" set option to 1, launch a bind and power on the brick you want to control. Repeat this for every brick using a different RX number each time and then indicate the number of bricks to be controlled using the Option field."

Git competely lost as a beginner : Where i to set (input) the RX number ?
Is there some kind of beginner howto somewhere , a step by step guide like the istruktion on updating firmware ?

**Comment by pascallanger** [comment]
Ok with a TX16S it should be quite simple. Select the MULDKG protocol, subprotocol either digital or analog depending of your bricks. The RX number appears on the TX16S as "Receiver number" which is by default set to 0. For option, you have a field "Option" or another name I can't remember which is at 0 by default and goes from -127 to +127, this is the one referenced here. It's all under where you select the protocol.

**Comment by marcrupprath** [comment]
It works very well with mouldking v4.0 4 channel boxes.
Do you have plans to implement protokoll of new 6 channel boxes , too ?

**Comment by marcrupprath** [comment]
Update about Mouldking 6 Channel boxes (not tested yet) : 
I have heard a story that it is possible to control V4.0 / 4.1 Boxes by using the Remote that comes with 6 Channel V6.0 Boxes.

Obviously it seesm that basic protocol is the same.
I ask myself how Channel 5 and 6 are handled.
When my 6 Channel Box, i have ordered for testing arives, i will test if it binds to RC Transmitter and how it behaves.

@pascallanger : What do you need for do a reference check ?
                          If you like i can send the 6 channel box with analog mouldking remote and 1 or 2 Motors for Testing

**Reply by pascallanger** [reply]
Is the 6 channel boxes/remotes analog or digital? Are all the channels analog or digital or a mix?
Looking quickly at the code:
 - The analog mode is sending 4 channels but there is space in the payload for 2 additional channels which are currently sent with a 0 value. A total of 6 analog channels could be sent without modifying the protocol apart from the current 0 values which would make the motors spin at full speed but they could filter that specific value. If they are using this available space, a 4 channel remote could control a 6 channel box and a 6 channel remote could control a 4 channel box. So this will be a good test when you receive your set.
 - The digital mode is sending 4 digital sticks (A, B, C, D) and 2 buttons (E,F). The buttons are commented in my code so not sure what they where used for... These 2 buttons could also be replaced by 2 digital sticks so a total of 6 digital channels could be sent without modifying the protocol. Same as above, a 4 channel remote could control a 6 channel box and a 6 channel remote could control a 4 channel box.
 - They could also have added a different header for the additional channels. Therefore the protocol will stay compatible, it's just that the box will ignore the new header.

So let me know when you have your set and do the cross checks. We might be able to test a few things remotely before shipping hardware.

**Comment by marcrupprath** [comment]
Updates about my test:
a)
I am not quite shure if all 6 cjannels in the box can be used as analog channels, because only the control elements for the first 4 channels are implemented as gimbels in the original remote, the others are buttons.
But the battery box looks very symetric . So i expect that all 6 channels can be alanog.

b)
The 6 Channel remote was able  to control a 4 channel box . But i was unable to bind the 4 channel remote to 6 channel box.

c) Very promissing : 
I was able to bind my Jumper Lite Remote (latest Edge TX and 4 in 1 version) to the 6 Channel Battery box and can control
4 of 6 channels as you expect ;-).


Next steps:
I would like two have an updated version of 4in1 firmware with changes you described.
Then i will check analog as well as digital options and report ;-)

**Reply by pascallanger** [reply]
> I was able to bind my Jumper Lite Remote (latest Edge TX and 4 in 1 version) to the 6 Channel Battery box and can control
4 of 6 channels as you expect ;-)

Did you use Analog or Digital to control the 6 channel box? Are both working with 6 channel box or only one?

**Comment by marcrupprath** [comment]
Good Morning ,
made short cross check

I can bind the Jumper Remote to Mouldking 6.0 , 6 Channelbox only in Analog mode.
I was unable to bind to this box in digital mode

Hope this helps ;.)

Have a nice day 
Marc

**Comment by marcrupprath** [comment]
Question about older 4 Channel boxes
I have noticed something different , see this kit ; Mould King Allrad-Buggy (18002)
I can bind to this older box in digital mode (unable in analog mode ) but obviously nothing moves.
Have checekd with mouldking handset and it works.

Something i have to keep in mind about it ? What may be the reason for this behaiviour ? 


Note:
They must have changed something in the protocol between these boxes i think ?!

But in the end it is not much important, at least i will change older box to V4.0 onedce that is working fine ,
Just to know ;-)

**Reply by pascallanger** [reply]
I'm a little confused on what old/new is... The page you pointed states that the box is V4.0 in the kit and you say that you'll upgrade to v4.0 so...
You are using an external module, correct? If that's the case we could work out what is being sent to these "older" boxes but also what's being sent to the 6 channels instead of guessing.

**Comment by marcrupprath** [comment]
Sorry for confusing you , ignore comment of "old" V4 box, because i have a work around for it ;-)
I will simple replace it with the "working" Version.
It seesm that there are different versions of the 4 Channel Boxes on the market.
Because the (older) Boxes, will also not work with the actual Mobile phone App, even both of them are 4 Channels remote.

By the way: All my Open TX/Exge TX RC handsets using internal 4-1 Modules,
For reference and additional testing ,primary  for planned use my own future Transmitter projects, i have external 4-1 modules in addition.


It is more important to me if you stay focused on upgrading support for New 6 Channel boxes in MouldKing Protocol 
My feeling / conclusion here :

a) The Mouldking button only remotes are unable  to bind to new V6.0 - 6 Channel analog box.
    The bind succesfully to my two different V4 Boxes (the version you have tested your development with and the "strange older" ones)

b) The New MouldkingV6 Handsets with Gimbals and Buttons can bind to "your" V4.0 Boxes as well to new V6.0 Boxes.
    To me it is unimporttant if the may bind with the "old series" of V4 boxes, have not tested. 
    
c) At least my Jumper T Lite with internal 4-1 module, sucessfully bind  in Analog mode  V6.0 as well as  actual V4.0 Boxes  
   I have never tested how it beavious in digital mode and actual  V4 Boxes because i don't need, i prefer Anallog mode whenever possible 
  to be s flexible as possible. 
   The Jumper does not bind to V6 Box in digital Mode.
   The Jumper do bind to "old" V4 boxes in digital Mode but was unable to control an output for unknown reasons.

My Solution about confusion of different Versions of 4 Channel Boxes:
You may get an Model with these "outdated" boxes if you by old Mouldking sets, from used market or "special offers"
The boxes can be simply replaced in most cases. The V4.0 boxes that can be currently bought "as new"  are alway the version that runs smoothly with your code .  Because the replacement does not cost much it is ok for me.

So according to the 4 Channel boxes: No need for any further action i would say . 


Hopes that helps and does not confuse you more than before ;-)

**Reply by pascallanger** [reply]
Please try the latest test version https://nightly.link/pascallanger/DIY-Multiprotocol-TX-Module/workflows/main/master/multi-test-build.zip
Binding only 1 box, see if you get control over the 6 channels.

**Comment by marcrupprath** [comment]
Update: I was able to control all 6 channels  in analog mode by using your provided Testversion

**Reply by pascallanger** [reply]
What happens when you go to -100% on channel 5 or 6?

**Reply by pascallanger** [reply]
Since it works, I could change the code to control with:
 - CH1..6: 1 analog 6 or 4 box sending 6 channels
 - CH7..12: 1 analog 6 or 4 box sending 6 channels
 - CH13..16: 1 analog 6 or 4 box sending 4 channels

Which makes a total of 3 boxes using 6+6+4=16 channels supported by the module.
Does it make sense?

**Comment by marcrupprath** [comment]
a) I Thought i did , but will cross check -100 Settings of channels 5+6

Your Idea makes sende for this specific configuration: 6+6+4,
because in all other cases te user must calculte the channel offset, because he can't "reach" the last 2 channels if receiver is a 4 channelbox, right.

So your idea makes sense exscpecialy for usthers that prefer the 6 channel boxes .
The can use up to 2 Boxes in a row and for the third box they are free if choice using a 6 channel box, leaving highest channels unused, or use a 4 channel box.

This should written in the comment (Protocol help file) and release notes.

What do you think about this:
We have one Mouldkingprotocol as Before (4 Channel Boxes only) and rename it as: "MouldKg_4444", or leaving the original name
And left a comment in release notes nd helpfile that zhis is for 4 channel boxes only.
This protocol has analog and digital subs , untouched as before.

The idea you described is put into an additioal Protocol , and we name it as : "MouldKg_664"
This protocoll is analog only to keep it simple bevause it it primary intended to use with 6 channel boxes.
Idea behind is that suffixes "4444" and"664" indicating the allowed (useful) configurations here 

What do you think ?

**Reply by pascallanger** [reply]
Can you also check the channels order? In the current code I had to swap the channels to match the ABCD outputs. I just want to be sure that CH5 is giving you E and CH6 F.
I'll more create a new subprotocol and rename the current one(s):
 - option 1: protocol MouldKg subprotocols: Analog4, Digital4, Analog6
 - option 2: protocol MouldKg subprotocols: A4444, D4444, A664

Let me know which one you would prefer, I think option1 is easier as e new user might not understand what 4444 and 664 mean but...

**Reply by marcrupprath** [reply]
I prefer option 2 with some aditional description in the helpfile.
Today i will carefully re check chanel orders A-F behaviour between -100 <- >100

**Reply by pascallanger** [reply]
Too late, I've implemented option 1 which I can change but with this new test version v1.3.4.45 you can test to control multiple 6 ports bricks.
https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/Protocols_Details.md#mouldkg---90

**Reply by pascallanger** [reply]
If you download the latest test version, it will have the option 2 names for the subprotocols.
Documentation updated too: https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/Protocols_Details.md#mouldkg---90

**Comment by marcrupprath** [comment]
Positive: Works like a charm and documentation is ok and readable i think

Have tested all 6 Channels according to : -100 <-> 100 , all OK
Except on minor thing,: Currently CH5 goes to  F  and CH6  to E  ;-)

**Reply by pascallanger** [reply]
Ok I'll swap that in the next version. 
Have you tried to control 2 and 3 bricks at the same time?

**Comment by marcrupprath** [comment]
Currently i only have 2 4channel Boxes and one 6 channel box.
I used both 4 channel boxes and tested all 8 channels here, all fine ;-)

When i got my other model i can do at least the 4444 Test.
Will simulate the 664 chain with 644 config ;-)

**Reply by pascallanger** [reply]
New firmware available with channels swap to match E and F. With your current boxes/bricks, you could test 644, 464 and 444.

If you want to ship me one of the old controller and brick, I could add them so you don't need to replace them. It shouldn't be too difficult...

**Comment by marcrupprath** [comment]
Testupdate:

a) My Digital 4 Channel box was nearly discharged last time , it works now as expected ;-)
b) Tested A4444 configuration with motors at Channels ;1,5,9,13 - all working as expected.
c) Tested A644  configuration with motors at channels:1,7,13 - 

   A problem ?
   Box #2 was bound as Receiver 2, Option 1 . Works in Single mode but in Test onle Box #1 with 6 Cjannels and #3 with 4 channels works.
   
   As i told you it is currently a minor problem too me, will retest when i have received othe 6 Cjannel box end of May

**Reply by pascallanger** [reply]
Sorry I don't understand the issue... Do you mean that in your test c 644, box 1 and 3 worked but not box 2?

**Reply by marcrupprath** [reply]
Yes that is what i have noticed.
But not so impotent right now, i will craoos check when i got my other 6 chanel boxe end of May

**Reply by pascallanger** [reply]
Hummm I need to check the code since this is unexpected. Basically if you test A4444, you should be able to just switch the sub protocol to A664 and still maintain connectivity to the first 3 boxes with the caveat that channels are moved up for the 2 last boxes.

**Comment by marcrupprath** [comment]
Small test result note about new Mouldking MK 3integrated Power module: It works fine with D4444 sub protocol.
Please add a note   in next Update of Documentation, thank you ;-)

**Reply by pascallanger** [reply]
What should I add? Is "Mouldking MK 3integrated Power module" a new product? Is there a name of reference for that module/brick?

**Reply by marcrupprath** [reply]
Hello, Yes the power module is quite new and used in these sets:
https://klemmbaustein.com/mould-king-15071/ (yellow block in image)
or
https://klemmbaustein.com/mould-king-17051-pistenraupe/ (as grey block)
Technical the same integrated system

It has 3 build in motors fpr Channels 1,2,3, 4 channels ist implemented by sing2 small sockets  to plug in 2 micro motors in parallel , assigned to channel 4

**Comment by marcrupprath** [comment]
Pascal: I would like to say thank you for the great work so far ;-)

Currently i am busy building some Mould King Kits for demonstrations and investigations about best practice according to "user experience" and usability.

Have cerated first demo videos on Youtube: https://www.youtube.com/watch?v=bS9ohuruMeA&t=123s

1. It seams that this possible was quite unknown in Brickset community ;-)
2. Results and commenst in other Youtube channels and what i received by Mail is very promissing ;-)
To make ling story short; Many Thanks, so far.

Because i am currently busy i must skip rest of multi box testings with number > 2 Boxes a litte bit ;-)

**Reply by pascallanger** [reply]
I think what's great about my implementation is that you don't have to rebind the boxes each time and you can control up to 4 at the same time.

**Comment by marcrupprath** [comment]
Testresult for latest test-build:1.3.4.50

A4444 - tested with all combinations from 1 - 4 boxes, works as expected
D4444-tested with all combinations from 1 - 4 boxes, works as expected
A664 - works as expected 
A64   - works as expexted
A644 - works as expected with respect to the channel offset for last box (starts with channel #13)

Very well done Pascal ;-)

**Comment by marcrupprath** [comment]
A question about "protocol invalid message" abbearing sometimes after updating the Multi Module.:
I have Jumper T16 With Edgetx 2.11 wunning and I started with latest stable Version : 1.3.4.31 
I have selected MouldKing in an Module profile, lets say in Analog sub profile.

After Updating to latest Tstversion 1.3.4.50 i got message: Protocoll invalid.

I have noticed that Mould King is not selectable s protocoll even if i have copied new text file in root folder of sc card.

What may causes this behaviour and how to force the "read in" of all the proocolls after updateing multi module ?

Thanks in advance ;-)

**Reply by pascallanger** [reply]
Invalid protocol means that the protocol is not available in the firmware you installed.
Mouldikng is part of the SFC automatic build but not AIR.

**Reply by marcrupprath** [reply]
Many thanks for this important clarification ;-)
I thought mouldking where present except in LBT versions !
Keep up good work !

**Comment by marcrupprath** [comment]
Hello Pascal,
i hope you are well, because it has been a while when we wjhere in last contact-
I am a little bit confused_
Because a daily testbuild seams to be no longer available as the well known zip collection i tried to compile the source by myself.

I was able to build a plattform io project to compile and flash bin by using the internal bootloader function of radiomaster tx 16 .
What confuses me ist the following:

The sub menue for mouldking provides only analog and digital = the view before you implemented the 6 ch box.

I asked myself why is it and how to chaange.
I have taken source code from here_

https://github.com/pascallanger/DIY-Multiprotocol-TX-Module
Is this the wrong place that does not contains all the changes of the latest test build ?

Thanks in advance

Marc
